### PR TITLE
Address #1058: Document scalac compiler options.

### PR DESCRIPTION
**Pull request by exoego**
This PR addresses #1058.

The metadata (`YAML`) and style (`HTML`/`Markdown`/`SCSS`) are separated, so that one can change look-and-feel easily.  

Almost all descriptions are extracted from [source](https://github.com/scala/scala/tree/1c56f0af6d3d59b7d2a8dbcf64077b0c1fe90f07/src/compiler/scala/tools/nsc/settings) and manually edited.

### CODE REVIEW

File: _data/compiler-options.yml [hunk 1]
```diff
@@ -0,0 +1,1011 @@
...
+          description: "Inspect both user-written code and expanded trees when generating unused symbol warnings."
+    - option: -Ywarn-dead-code
+      schema:
+        type: Boolean
+      description: "Warn when dead code is identified."
+    - option: -Ywarn-value-discard
+      schema:
+        type: Boolean
+      description: "Warn when non-Unit expression results are unused."
+      note: "To reduce noises for common idiom, warning are suppressed if type of results are `this.type`."
```
Comment (exoego) [inline _data/compiler-options.yml]: This note is for https://github.com/scala/bug/issues/10761

File: _overviews/compiler-options/index.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+layout: singlepage-overview
+title: Scala Compiler Options
+
+discourse: true
+---
+<style type="text/css">
+.definition-list dd { 
+    margin-left: 2em;
+}
+.definition-list .deprecated {
```
Comment (exoego) [inline _overviews/compiler-options/index.md]: Hope someone polish these `deprecated` and `note` styles.

File: _data/compiler-options.yml [hunk 3]
```diff
@@ -0,0 +1,1011 @@
...
+
+
+
+- category: Private Settings
+  description: |
+    Options with `-Y` prefix are more experimental and unstable than those with `-X` prefix.
+
+    Those influencing to language semantics are likely to be deprecated in Scala 2.13 (see [scala/scala-dev#430](https://github.com/scala/scala-dev/issues/430)).
+
+    Some optimization-related options, e.g. `-opt:**` and `-opt-inline-form` are included in this section, since **REASON_HERE**.
```
Comment (exoego) [inline _data/compiler-options.yml]: I found  `-opt:**` and `-opt-inline-form` are in private settings (-ish) section for `-Y' in [the source](https://github.com/scala/scala/blob/1c56f0af6d3d59b7d2a8dbcf64077b0c1fe90f07/src/compiler/scala/tools/nsc/settings/ScalaSettings.scala#L297) therefore put them in this section. 

Is it OK to categorize `-opt:**` and `-opt-inline-form` as private settings?

Comment (jvican) [reply _data/compiler-options.yml]: yes I think so :smile:

**Comment by jvican** [comment]
Wow, this is amazing work @exoego. How did you generate the yml file? Can that be done for future versions of Scala?

**Comment by exoego** [comment]
@jvican 

I used [a generator added in my fork of scala](
https://github.com/exoego/scala/blob/8bb426b4a655202d26bdb6b3529ea35f8418b8dc/src/exporter/scala/tools/nsc/ScalaCompilerOptionsExporter.scala), which can be executed via `sbt exporter/run`.
It generetates YAML from the instance of `scala.tools.nsc.Settings` via reflection.

Manual modification of several minutes are required, since the YAML file can contains extra useful information (such as why `-Ywarn-value-discard` does not warn on `this.type`) which are not obtained from the source code.

The generator is super quickie and dirty... 
I will clean up under guidance with my pleasure 😄

File: _data/compiler-options.yml (hunk 3 above)
Comment (NthPortal) [reply _data/compiler-options.yml]: I think that should say `-opt-inline-from`, not `-form`

Comment (exoego) [reply _data/compiler-options.yml]: Added reason and fixed typo.

**Comment by SethTisue** [comment]
(crazy busy lately, but I'll get to this)

**Comment by som-snytt** [comment]
Someone needed this today. Also, `-opt` might need its own wiki...

**Comment by som-snytt** [comment]
I don't know what the hang-up is, but someone needs this literally every day.

https://gitter.im/scala/scala?at=5b2b98d80168e70c08e067dc

**Comment by SethTisue** [comment]
apologies. I should have merged this first and worried about maybe improving it further later.

thank you, @exoego! this looks great already.

**Comment by SethTisue** [comment]
live at https://docs.scala-lang.org/overviews/compiler-options/index.html

**Comment by som-snytt** [comment]
That is a huge improvement over nothing! Of course, it's already out of date. But that just gives folks an excuse to head over to StackOverflow and ask a question and earn five points. In fact, I'm going over there now to ask what is `-Xmixin-force-forwarders` and when would I want to use it and does it have anything to do with World Cup?

**Comment by som-snytt** [comment]
https://stackoverflow.com/a/18603148/1296806

**Comment by SethTisue** [comment]
publicized: https://twitter.com/scala_lang/status/1011334202883747844

**Pull request merged**